src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -38,189 +38,109 @@
 #include "utilities/hashtable.inline.hpp"
 #include "utilities/macros.hpp"
 
-oop JvmtiTagMapEntry::object() {
-  return literal().resolve();
-}
 
-oop JvmtiTagMapEntry::object_no_keepalive() {
-  // Just peek at the object without keeping it alive.
-  return literal().peek();
-}
+JvmtiTagMapKey::JvmtiTagMapKey(oop obj) : _obj(obj) {}
 
-JvmtiTagMapTable::JvmtiTagMapTable()
-  : Hashtable<WeakHandle, mtServiceability>(_table_size, sizeof(JvmtiTagMapEntry)) {}
+JvmtiTagMapKey::JvmtiTagMapKey(const JvmtiTagMapKey& src) {
+  // move object into WeakHandle when copying into the table
+  assert(src._obj != nullptr, "must be set");
+  _wh = WeakHandle(JvmtiExport::weak_tag_storage(), src._obj);
+  _obj = nullptr;
+}
 
-void JvmtiTagMapTable::clear() {
-  // Clear this table
-  log_debug(jvmti, table)("JvmtiTagMapTable cleared");
-  for (int i = 0; i < table_size(); ++i) {
-    for (JvmtiTagMapEntry* m = bucket(i); m != NULL;) {
-      JvmtiTagMapEntry* entry = m;
-      // read next before freeing.
-      m = m->next();
-      free_entry(entry);
-    }
-    JvmtiTagMapEntry** p = bucket_addr(i);
-    *p = NULL; // clear out buckets.
+JvmtiTagMapKey::~JvmtiTagMapKey() {
+  // If obj is set null it out, this is called for stack object on lookup,
+  // and it should not have a WeakHandle created for it yet.
+  if (_obj != nullptr) {
+    _obj = nullptr;
+    assert(_wh.is_null(), "WeakHandle should be null");
+  } else {
+    _wh.release(JvmtiExport::weak_tag_storage());
   }
-  assert(number_of_entries() == 0, "should have removed all entries");
 }
 
-JvmtiTagMapTable::~JvmtiTagMapTable() {
-  clear();
-  // base class ~BasicHashtable deallocates the buckets.
+oop JvmtiTagMapKey::object() const {
+  assert(_obj == nullptr, "Must have a handle and not object");
+  return _wh.resolve();
 }
 
-// Entries are C_Heap allocated
-JvmtiTagMapEntry* JvmtiTagMapTable::new_entry(unsigned int hash, WeakHandle w, jlong tag) {
-  JvmtiTagMapEntry* entry = (JvmtiTagMapEntry*)Hashtable<WeakHandle, mtServiceability>::new_entry(hash, w);
-  entry->set_tag(tag);
-  return entry;
+oop JvmtiTagMapKey::object_no_keepalive() const {
+  assert(_obj == nullptr, "Must have a handle and not object");
+  return _wh.peek();
 }
 
-void JvmtiTagMapTable::free_entry(JvmtiTagMapEntry* entry) {
-  entry->literal().release(JvmtiExport::weak_tag_storage()); // release to OopStorage
-  BasicHashtable<mtServiceability>::free_entry(entry);
-}
+JvmtiTagMapTable::JvmtiTagMapTable() : _table(Constants::_table_size) {}
 
-unsigned int JvmtiTagMapTable::compute_hash(oop obj) {
-  assert(obj != NULL, "obj is null");
-  return obj->identity_hash();
-}
-
-JvmtiTagMapEntry* JvmtiTagMapTable::find(int index, unsigned int hash, oop obj) {
-  assert(obj != NULL, "Cannot search for a NULL object");
+void JvmtiTagMapTable::clear() {
+  struct RemoveAll {
+    bool do_entry(const JvmtiTagMapKey& entry, const jlong& tag) {
+      return true;
+    }
+  } remove_all;
+  _table.unlink(&remove_all);
```

Comment:
This logic is a bit unclear - can you add some comments. I guess I need to understand what unlink does when reading this code.